tests/sdk/test_parameter.py

```diff
@@ -0,0 +1,136 @@
+from parameterized import parameterized, param
+
+import pytest
+from unittest import TestCase
+from samtranslator.sdk.parameter import SamParameterValues
+from mock import patch
+
+class TestSAMParameterValues(TestCase):
+
+    def test_add_default_parameter_values_must_merge(self):
+        parameter_values = {
+            "Param1": "value1"
+        }
+
+        sam_template = {
+            "Parameters": {
+                "Param2": {
+                    "Type": "String",
+                    "Default": "template default"
+                }
+            }
+        }
+
+        expected = {
+            "Param1": "value1",
+            "Param2": "template default"
+        }
+
+        sam_parameter_values = SamParameterValues(parameter_values)
+        sam_parameter_values.add_default_parameter_values(sam_template)
+        self.assertEqual(expected, sam_parameter_values.parameter_values)
+
+    def test_add_default_parameter_values_must_override_user_specified_values(self):
+        parameter_values = {
+            "Param1": "value1"
+        }
+
+        sam_template = {
+            "Parameters": {
+                "Param1": {
+                    "Type": "String",
+                    "Default": "template default"
+                }
+            }
+        }
+
+        expected = {
+            "Param1": "value1"
+        }
+
+        sam_parameter_values = SamParameterValues(parameter_values)
+        sam_parameter_values.add_default_parameter_values(sam_template)
+        self.assertEqual(expected, sam_parameter_values.parameter_values)
+
+    def test_add_default_parameter_values_must_skip_params_without_defaults(self):
```

Comment:
This would fail CFN execution, right? All Params require a value. Can (should) we fail early? I like the idea of SAM failing faster than CFN when possible. Then again, it could be possible that a Plugin could fill in Params, but I can't think of a use-case. I'd like to get James and Keeton's feedback too.